test/unit/math/mix/fun/complex_schur_decompose_test.cpp

```diff
@@ -0,0 +1,94 @@
+#include <test/unit/math/test_ad.hpp>
+#include <stdexcept>
+
+TEST(mathMixFun, complexSchurDecomposeT) {
+  auto f = [](const auto& x) {
+    using stan::math::complex_schur_decompose_t;
+    return complex_schur_decompose_t(x);
+  };
+  for (const auto& x : stan::test::square_test_matrices(0, 2)) {
+    stan::test::expect_ad(f, x);
+  }
+
+  Eigen::MatrixXd a32(3, 2);
+  a32 << 3, -5, 7, -7.2, 9.1, -6.3;
+  EXPECT_THROW(f(a32), std::invalid_argument);
+}
+
+TEST(mathMixFun, complexSchurDecomposeU) {
+  auto f = [](const auto& x) {
+    using stan::math::complex_schur_decompose_u;
+    return complex_schur_decompose_u(x);
+  };
+  for (const auto& x : stan::test::square_test_matrices(0, 2)) {
+    stan::test::expect_ad(f, x);
+  }
+
+  Eigen::MatrixXd a32(3, 2);
+  a32 << 3, -5, 7, -7.2, 9.1, -6.3;
+  EXPECT_THROW(f(a32), std::invalid_argument);
+}
+
+template <typename V>
+void test_complex_schur_decompose(const Eigen::MatrixXd& x) {
+  using stan::math::complex_schur_decompose_t;
+  using stan::math::complex_schur_decompose_u;
+  using stan::math::value_of_rec;
+  Eigen::Matrix<V, -1, -1> X(x.rows(), x.cols());
+  for (int i = 0; i < x.size(); ++i)
+    X(i) = x(i);
+  auto T = complex_schur_decompose_t(X);
+  auto U = complex_schur_decompose_u(X);
+  auto X2 = U * T * U.adjoint();
+  for (int j = 0; j < x.cols(); ++j) {
+    for (int i = 0; i < x.rows(); ++i) {
+      EXPECT_FLOAT_EQ(x(i, j), value_of_rec(X2(i, j).real()));
+    }
+  }
+}
+
+template <typename V>
+void test_complex_schur_decompose_complex(const Eigen::MatrixXd& x) {
+  using stan::math::complex_schur_decompose_t;
+  using stan::math::complex_schur_decompose_u;
+  using stan::math::value_of_rec;
+  Eigen::Matrix<std::complex<V>, -1, -1> X(x.rows(), x.cols());
+  for (int i = 0; i < x.size(); ++i)
+    X(i) = std::complex<double>(x(i), i);
+  auto T = complex_schur_decompose_t(X);
+  auto U = complex_schur_decompose_u(X);
+  auto X2 = U * T * U.adjoint();
+  for (int j = 0; j < x.cols(); ++j) {
+    for (int i = 0; i < x.rows(); ++i) {
+      EXPECT_NEAR(value_of_rec(X(i, j).real()), value_of_rec(X2(i, j).real()),
+                  1e-8);
+      EXPECT_NEAR(value_of_rec(X(i, j)).imag(), value_of_rec(X2(i, j).imag()),
+                  1e-8);
+    }
+  }
+}
+
+TEST(mathMixFun, complexSchurDecompose) {
```

Comment:
What is this adding to the `test_ad` function?  That already tests values, gradients, and exception behavior for all the arguments.

Reply:
Copied from old tests which didn't have the `prim` tests at all, so I believe these were the only tests that the function actually satisfied the required properties. Do you want me to remove them?

Reply:
My first thought was to move these into `prim` without the extra types (so just using square_test_matrices to test double matrices of different sizes) but it turns out that this function is in the `test_ad` header rather than the normal `utils` header